File: app/src/main/java/com.github.anrimian.musicplayer/ui/utils/views/recycler_view/touch_helper/drag_and_drop/SimpleItemTouchHelperCallback.java
package com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.drag_and_drop;


import android.animation.ObjectAnimator;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public class SimpleItemTouchHelperCallback extends ItemTouchHelper.Callback{

    private static final float BOTTOM_Z = 0f;
    private static final float TOP_Z = 8f;

    private final boolean horizontalDrag;
    private final boolean isLongPressDragEnabled;

    private OnMovedListener onMovedListener;
    private OnStartDragListener onStartDragListener;
    private OnEndDragListener onEndDragListener;

    public SimpleItemTouchHelperCallback(boolean isLongPressDragEnabled) {
        this(false, isLongPressDragEnabled);
    }

    public SimpleItemTouchHelperCallback(boolean horizontalDrag, boolean isLongPressDragEnabled) {
        this.horizontalDrag = horizontalDrag;
        this.isLongPressDragEnabled = isLongPressDragEnabled;
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        if (onStartDragListener != null && actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
            onStartDragListener.onStartDrag();
        }
        setIsDragging(viewHolder, true);
        super.onSelectedChanged(viewHolder, actionState);
    }

    @Override
    public void clearView(@NonNull RecyclerView recyclerView,
                          @NonNull RecyclerView.ViewHolder viewHolder) {
        if (onEndDragListener != null) {
            onEndDragListener.onEndDrag();
        }
        setIsDragging(viewHolder, false);
        super.clearView(recyclerView, viewHolder);
    }

    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView,
                          @NonNull RecyclerView.ViewHolder viewHolder,
                          @NonNull RecyclerView.ViewHolder target) {
        if (onMovedListener != null) {
            onMovedListener.onItemMoved(viewHolder.getBindingAdapterPosition(), target.getBindingAdapterPosition());
            return true;
        }
        return false;
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {

    }

    @Override
    public int getMovementFlags(@NonNull RecyclerView recyclerView,
                                @NonNull RecyclerView.ViewHolder viewHolder) {
        int directions = ItemTouchHelper.DOWN | ItemTouchHelper.UP;
        if (horizontalDrag) {
            directions |= ItemTouchHelper.START | ItemTouchHelper.END;
        }
        return makeFlag(ItemTouchHelper.ACTION_STATE_DRAG, directions);
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return isLongPressDragEnabled;
    }

    public void setOnMovedListener(OnMovedListener onMovedListener) {
        this.onMovedListener = onMovedListener;
    }

    public void setOnStartDragListener(OnStartDragListener onStartDragListener) {
        this.onStartDragListener = onStartDragListener;
    }

    public void setOnEndDragListener(OnEndDragListener onEndDragListener) {
        this.onEndDragListener = onEndDragListener;
    }

    private void setIsDragging(RecyclerView.ViewHolder viewHolder, boolean dragging) {
        if (viewHolder == null) {
            return;
        }
        if (viewHolder instanceof DragListener) {
            ((DragListener) viewHolder).onDragStateChanged(dragging);
        }
        float baseElevation = dragging? BOTTOM_Z: TOP_Z;
        float dragElevation = dragging? TOP_Z: BOTTOM_Z;
        ObjectAnimator elevationAnimator = ObjectAnimator.ofFloat(viewHolder.itemView,
                "translationZ",
                baseElevation,
                dragElevation);
        elevationAnimator.setDuration(DEFAULT_DRAG_ANIMATION_DURATION);
        elevationAnimator.start();
    }

    public interface OnMovedListener {
        void onItemMoved(int from, int to);
    }

    public interface OnStartDragListener {
        void onStartDrag();
    }

    public interface OnEndDragListener {
        void onEndDrag();
    }

}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/utils/views/recycler_view/touch_helper/drag_and_drop/SimpleItemTouchHelperCallback.kt
package com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.drag_and_drop

import android.animation.ObjectAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

class SimpleItemTouchHelperCallback(
    private val horizontalDrag: Boolean = false,
    private val isLongPressDragEnabled: Boolean = true,
    private val shouldNotDragViewHolder: (RecyclerView.ViewHolder) -> Boolean = { false },
    private val dragElevation: Float = TOP_Z
) : ItemTouchHelper.Callback() {

    private var onMovedListener: ((from: Int, to: Int) -> Unit)? = null
    private var onStartDragListener: ((position: Int) -> Unit)? = null
    private var onEndDragListener: ((position: Int) -> Unit)? = null

    private var startDragX = 0f
    private var startDragY = 0f

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        if (onStartDragListener != null && actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
            onStartDragListener!!(viewHolder!!.bindingAdapterPosition)
            startDragX = viewHolder.itemView.x
            startDragY = viewHolder.itemView.y
        }
        setIsDragging(viewHolder, true)
        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        if (onEndDragListener != null) {
            onEndDragListener!!(viewHolder.bindingAdapterPosition)
        }
        setIsDragging(viewHolder, false)
        super.clearView(recyclerView, viewHolder)
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder,
    ): Boolean {
        if (shouldNotDragViewHolder(target)) {
            return false
        }
        if (onMovedListener != null) {
            onMovedListener!!(viewHolder.bindingAdapterPosition, target.bindingAdapterPosition)
            return true
        }
        return false
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
    ): Int {
        if (shouldNotDragViewHolder(viewHolder)) {
            return ItemTouchHelper.ACTION_STATE_IDLE
        }
        var directions = ItemTouchHelper.DOWN or ItemTouchHelper.UP
        if (horizontalDrag) {
            directions = directions or (ItemTouchHelper.START or ItemTouchHelper.END)
        }
        return makeFlag(ItemTouchHelper.ACTION_STATE_DRAG, directions)
    }

    override fun isLongPressDragEnabled(): Boolean {
        return isLongPressDragEnabled
    }

    fun setOnMovedListener(onMovedListener: (from: Int, to: Int) -> Unit) {
        this.onMovedListener = onMovedListener
    }

    fun setOnStartDragListener(onStartDragListener: (position: Int) -> Unit) {
        this.onStartDragListener = onStartDragListener
    }

    fun setOnEndDragListener(onEndDragListener: (position: Int) -> Unit) {
        this.onEndDragListener = onEndDragListener
    }

    private fun setIsDragging(viewHolder: RecyclerView.ViewHolder?, dragging: Boolean) {
        if (viewHolder == null) {
            return
        }
        if (viewHolder is DragListener) {
            viewHolder.onDragStateChanged(dragging)
        }
        val baseElevation = if (dragging) BOTTOM_Z else dragElevation
        val dragElevation = if (dragging) dragElevation else BOTTOM_Z
        val elevationAnimator = ObjectAnimator.ofFloat(
            viewHolder.itemView,
            "translationZ",
            baseElevation,
            dragElevation
        )
        elevationAnimator.duration = DEFAULT_DRAG_ANIMATION_DURATION.toLong()
        elevationAnimator.start()
    }

    companion object {
        private const val BOTTOM_Z = 0f
        private const val TOP_Z = 8f
    }

}